### PR TITLE
Prmdr 355 rbac

### PR DESCRIPTION
**Pull request by SRAlexander**

### CODE REVIEW

File: lambdas/utils/exceptions.py
Comment (thisusernameisnowtaken) [inline lambdas/utils/exceptions.py:1]: Should this be UserNotAuthorisedException?

File: lambdas/handlers/authoriser_handler.py [hunk 1]
```diff
@@ -63,27 +61,54 @@ def lambda_handler(event, context):
     principal_id = ""
     _, _, _, region, aws_account_id, api_gateway_arn = event["methodArn"].split(":")
     api_id, stage, _http_verb, _resource_name = api_gateway_arn.split("/")
-
+    user_role = decoded["repository_role"]
```
Comment (NogaNHS) [inline lambdas/handlers/authoriser_handler.py:64]: Is there a reason this is outside the try-except? code there be a keyError here?

File: lambdas/handlers/search_patient_details_handler.py [hunk 2]
```diff
@@ -30,10 +36,27 @@ def lambda_handler(event, context):
 
     try:
         nhs_number = event["queryStringParameters"]["patientId"]
+        client = boto3.client("ssm")
+        ssm_public_key_parameter_name = os.environ["SSM_PARAM_JWT_TOKEN_PUBLIC_KEY"]
+        ssm_response = client.get_parameter(
+            Name=ssm_public_key_parameter_name, WithDecryption=True
+        )
+        public_key = ssm_response["Parameter"]["Value"]
```
Comment (NogaNHS) [inline lambdas/handlers/search_patient_details_handler.py:44]: Could we use the SSM service here?

File: lambdas/handlers/authoriser_handler.py (hunk 1 above)
Comment (SRAlexander) [reply lambdas/handlers/authoriser_handler.py:67]: We sure can, updated in next push.

File: lambdas/handlers/search_patient_details_handler.py (hunk 2 above)
Comment (SRAlexander) [reply lambdas/handlers/search_patient_details_handler.py:44]: We sure can, updated in next push.

File: lambdas/handlers/authoriser_handler.py [hunk 3]
```diff
@@ -18,26 +18,31 @@
...
-            Name=ssm_public_key_parameter_name, WithDecryption=True
-        )
-        public_key = ssm_response["Parameter"]["Value"]
+        ssm_service = SSMService()
+        ssm_public_key_parameter_name = os.environ["SSM_PARAM_JWT_TOKEN_PUBLIC_KEY"]
+        
+        token = event["authorizationToken"]
+        public_key  = ssm_service.get_ssm_parameter(ssm_public_key_parameter_name, True)
 
+        logger.info(public_key)
```
Comment (thisusernameisnowtaken) [inline lambdas/handlers/authoriser_handler.py:43]: Need to remember to remove these extra logs before merging with main

File: lambdas/handlers/search_patient_details_handler.py [hunk 4]
```diff
@@ -70,7 +92,7 @@ def lambda_handler(event, context):
     except UserNotAuthorisedException as e:
         logger.error(f"PDS not authorised patient: {str(e)}")
         return ApiGatewayResponse(
-            404, "Patient is outside of your access policy", "GET"
+            404, "Patient does not exist for given NHS number", "GET"
```
Comment (thisusernameisnowtaken) [inline lambdas/handlers/search_patient_details_handler.py:95]: Could this message be improved to reflect that the requestors ODS code may have prevented them seeing the record?

File: lambdas/handlers/token_handler.py [hunk 5]
```diff
@@ -57,34 +48,34 @@ def token_request(oidc_service, ods_api_service, event):
 
         logger.info("Fetching access token from OIDC Provider")
         access_token, id_token_claim_set = oidc_service.fetch_tokens(auth_code)
-        logger.info(f"id_token_claim_set:  {id_token_claim_set}")
-
+        
         logger.info(f"Access token: {access_token}")
```
Comment (thisusernameisnowtaken) [inline lambdas/handlers/token_handler.py:52]: Lots of logs to tidy up before merging with main

File: lambdas/handlers/token_handler.py [hunk 6]
```diff
@@ -114,117 +105,46 @@ def token_request(oidc_service, ods_api_service, event):
...
-#         return ApiGatewayResponse(
-#             500, "Failed to fetch organisation data from ODS", "GET"
-#         ).create_api_gateway_response()
-#
-#     return ApiGatewayResponse(
-#         200, json.dumps(response), "GET"
-#     ).create_api_gateway_response()
-
 
 def generate_repository_role(organisation: dict, smartcart_role: str):
```
Comment (thisusernameisnowtaken) [inline lambdas/handlers/token_handler.py:115]: A potential service method? Maybe save for later improvements once FoT is done

File: lambdas/handlers/token_handler.py [hunk 7]
```diff
@@ -262,19 +182,12 @@ def create_login_session(id_token_claim_set: IdTokenClaimSet) -> str:
 def issue_auth_token(
     session_id: str,
     id_token_claim_set: IdTokenClaimSet,
-    permitted_orgs_and_roles: list[dict],
+    permitted_orgs_and_role: list[dict],
```
Comment (thisusernameisnowtaken) [inline lambdas/handlers/token_handler.py:185]: user org details might be a better name for this field

File: lambdas/services/oidc_service.py [hunk 8]
```diff
@@ -96,8 +95,88 @@ def validate_and_decode_token(self, signed_token: str) -> Dict:
...
+            if role_code[0].upper() == prefix_character.upper():
+                return role_code
+
+        raise AuthorisationException(f'Role codes have been found for the user but not with prefix {prefix_character.upper()}')
 
-    # TODO AKH SSM service
+    def fetch_userinfo(self, access_token: AccessToken) -> Dict:
+        logger.info(f"Access token for user info request: {access_token}")
+
+        # params={"scope": "nationalrbacaccess"},
```
Comment (thisusernameisnowtaken) [inline lambdas/services/oidc_service.py:159]: remove this comment

**Review by thisusernameisnowtaken — APPROVED** [review]
Just some minor changes, primarily removing some of the logs that have been added to aid debugging

File: lambdas/handlers/search_patient_details_handler.py (hunk 4 above)
Comment (SRAlexander) [reply lambdas/handlers/search_patient_details_handler.py:95]: Negative, that information could prevent a security risk.

File: lambdas/handlers/token_handler.py (hunk 5 above)
Comment (SRAlexander) [reply lambdas/handlers/token_handler.py:52]: Resolved

File: lambdas/services/oidc_service.py (hunk 8 above)
Comment (SRAlexander) [reply lambdas/services/oidc_service.py:159]: resolved

File: lambdas/handlers/token_handler.py (hunk 6 above)
Comment (SRAlexander) [reply lambdas/handlers/token_handler.py:115]: service level indeed on the code tidy excerises

**Pull request merged**